db/write_thread.cc

```diff
@@ -310,137 +356,295 @@ size_t WriteThreadImpl::EnterAsBatchGroupLeader(
 }
 
-void WriteThreadImpl::ExitAsBatchGroupLeader(Writer* leader,
-                                             Writer* last_writer,
-                                             Status status) {
+void WriteThread::ExitAsBatchGroupLeader(WriteGroup& write_group,
+                                         Status status) {
+  static AdaptationContext ctx("ExitAsBatchGroupLeader");
+  Writer* leader = write_group.leader;
+  Writer* last_writer = write_group.last_writer;
   assert(leader->link_older == nullptr);
 
-  Writer* head = newest_writer_.load(std::memory_order_acquire);
-  if (head != last_writer ||
-      !newest_writer_.compare_exchange_strong(head, nullptr)) {
-    // Either w wasn't the head during the load(), or it was the head
-    // during the load() but somebody else pushed onto the list before
-    // we did the compare_exchange_strong (causing it to fail).  In the
-    // latter case compare_exchange_strong has the effect of re-reading
-    // its first param (head).  No need to retry a failing CAS, because
-    // only a departing leader (which we are at the moment) can remove
-    // nodes from the list.
-    assert(head != last_writer);
-
-    // After walking link_older starting from head (if not already done)
-    // we will be able to traverse w->link_newer below. This function
-    // can only be called from an active leader, only a leader can
-    // clear newest_writer_, we didn't, and only a clear newest_writer_
-    // could cause the next leader to start their work without a call
-    // to MarkJoined, so we can definitely conclude that no other leader
-    // work is going on here (with or without db mutex).
-    CreateMissingNewerLinks(head);
-    assert(last_writer->link_newer->link_older == last_writer);
-    last_writer->link_newer->link_older = nullptr;
-
-    // Next leader didn't self-identify, because newest_writer_ wasn't
-    // nullptr when they enqueued (we were definitely enqueued before them
-    // and are still in the list).  That means leader handoff occurs when
-    // we call MarkJoined
-    SetState(last_writer->link_newer, STATE_GROUP_LEADER);
-  }
-  // else nobody else was waiting, although there might already be a new
-  // leader now
-
-  while (last_writer != leader) {
-    last_writer->status = status;
-    // we need to read link_older before calling SetState, because as soon
-    // as it is marked committed the other thread's Await may return and
-    // deallocate the Writer.
-    auto next = last_writer->link_older;
-    SetState(last_writer, STATE_COMPLETED);
-
-    last_writer = next;
+  if (enable_pipelined_write_) {
+    Writer* next_leader = nullptr;
+    // We attempt to put a dummy writer to the head of newest_writer_ list
+    // with a CAS operation, to block the next batch group leader.
+    // If the CAS operation fail, we know the next leader is in the list at
+    // this point. We simply don't wake it up until we are ready.
```

Comment:
This is to avoid the next leader finish its work and join memtable writer queue before the current leader do. Will see if I can have another way that don't require dummy node.